src/sentry/utils/snuba.py

```diff
@@ -0,0 +1,162 @@
+    project_ids = list(set.intersection(*[set(ids) for ids in project_ids if ids]))
+
+    # If the grouping, aggregation, or any of the conditions reference `issue`
+    # we need to fetch the issue definitions (issue -> fingerprint hashes)
+    references_issues = 'issue' in groupby + [aggregateby] + [c[0] for c in conditions]
+    issues = get_project_issues(project_ids) if references_issues else None
+
+    url = '{0}/query'.format(SNUBA)
+    request = {k: v for k, v in six.iteritems({
+        'from_date': start.isoformat(),
+        'to_date': end.isoformat(),
+        'conditions': conditions,
+        'groupby': groupby,
+        'project': project_ids,
+        'aggregation': aggregation,
+        'aggregateby': aggregateby,
+        'granularity': rollup,
+        'issues': issues,
+        'arrayjoin': arrayjoin,
+    }) if v is not None}
+
+    response = requests.post(url, data=json.dumps(request))
+    # TODO handle error responses
+    response = json.loads(response.text)
+
+    # Validate and scrub response, and translate snuba keys back to IDs
+    expected_cols = groupby + ['aggregate']
+    assert all(c['name'] in expected_cols for c in response['meta'])
+    for d in response['data']:
+        if 'time' in d:
+            d['time'] = int(to_timestamp(parse_datetime(d['time'])))
+        if d['aggregate'] is None:
+            d['aggregate'] = 0
+        for col in rev_snuba_map:
+            if col in d:
+                d[col] = rev_snuba_map[col][d[col]]
+
+    return nest_groups(response['data'], groupby)
+
+
+def nest_groups(data, groups):
+    """
+    Build a nested mapping from query response rows. Each group column
+    gives a new level of nesting and the leaf result is the aggregate
+    """
+    if not groups:
+        # If no groups, just return the aggregate value from the first row
+        return data[0]['aggregate'] if data else None
+    else:
+        g, rest = groups[0], groups[1:]
+        inter = {}
+        for d in data:
+            inter.setdefault(d[g], []).append(d)
+        return {k: nest_groups(v, rest) for k, v in six.iteritems(inter)}
+
+# The following are functions for resolving information from sentry models
+# about projects, environments, and issues (groups). Having this snuba
+# implementation have to know about these relationships is not ideal, and
+# many of these relationships (eg environment id->name) will have already
+# been queried and exist somewhere in the call stack, but for now, lookup
+# is implemented here for simplicity.
+
+
+def get_snuba_map(column, ids):
+    """
+    Some models are stored differently in snuba, eg. as the environment
+    name instead of the the environment ID. Here we look up a set of keys
+    for a given model and return a lookup dictionary from those keys to the
+    equivalent ones in snuba.
+    """
+    mappings = {
+        'environment': (Environment, 'name'),
```

Comment:
nit: I'd prefer to call it an `environment_id` (and so on) when it is an ID, as we do elsewhere. I guess that'd need another level of translation to Snuba columns though, bleh.